web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php

```diff
@@ -26,47 +26,53 @@ class CustomWeather extends BlockBase implements ContainerFactoryPluginInterface
   protected string $apiKey;
 
   /**
-   * Stores the selected city.
+   * Stores current user city.
+   *
+   * @var mixed|string
    */
-  protected string $selectedCity;
+  protected mixed $city;
 
   /**
    * {@inheritdoc}
    */
-  public function __construct(protected ClientFactory $httpClient, array $configuration, $plugin_id, $plugin_definition, protected ConfigFactoryInterface $configFactory) {
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, protected ClientFactory $httpClient, protected ConfigFactoryInterface $configFactory, protected UserCityHandler $userCityHandler) {
     parent::__construct($configuration, $plugin_id, $plugin_definition);
     $this->apiKey = $this->configFactory->get('custom_weather.settings')->get('api_key');
-    $this->selectedCity = $this->configFactory->get('custom_weather.settings')->get('selected_city');
+    $this->city = $this->userCityHandler->getCurrentCity();
   }
 
   /**
    * {@inheritdoc}
    */
   public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition): static {
     return new static(
-      $container->get('http_client_factory'),
       $configuration,
       $plugin_id,
       $plugin_definition,
+      $container->get('http_client_factory'),
       $container->get('config.factory'),
+      $container->get('custom_weather.user_city_handler'),
     );
   }
 
   /**
    * {@inheritdoc}
    */
   public function build():array {
-    $weatherData = $this->getWeatherApi();
+    $apiKey = $this->apiKey;
+    $city = $this->city;
+    $httpClient = $this->httpClient;
+    $weatherData = $this->userCityHandler->getWeatherApi($apiKey, $city, $httpClient);
     if ($weatherData) {
       $temp = round($weatherData['main']['temp'] - 273.15, 1);
       $weather_text = $weatherData['weather'][0]['main'];
-      $selected_city = $this->configFactory->get('custom_weather.settings')->get('selected_city');
       return [
         '#theme' => 'custom_weather_block',
         '#temp' => $temp,
         '#weather_text' => $weather_text,
-        '#selected_city' => $selected_city,
+        '#selected_city' => $this->city,
```

Comment:
Another example why you are not doing much by assigning $this->$property to variable directly
You assigned $city before, but here you are using $this->city
These assignments are useless